### PR TITLE
Simple rewiring of spliced specs

### PR DESCRIPTION
**Pull request by nhanford**
Rewiring is an operation on concrete, installed, spliced specs which carefully relocates any relevant prefixes throughout all the specs in a spliced DAG and then performs necessary operations for the spliced spec to be seen as "installed." This operation remains somewhat "dangerous" and can lead to ABI breaks and other unexpected behavior. It should be used with caution.
Rewiring extends splicing (#20262) and honors all the semantics therein. For example, when a transitive splice is rewired, it will bring in all the dependencies of the newly-spliced dependency. Like splicing, rewiring is not destructive--it creates a new spec that conforms to the new spec from the splice.

### CODE REVIEW

File: lib/spack/spack/rewiring.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+                                           str(spack.store.layout.root),
+                                           prefix_to_prefix,
+                                           False,
+                                           spec.build_spec.prefix,
+                                           spec.prefix)
+        relocate.relocate_text_bin(binaries=bins_to_relocate,
+                                   prefixes=prefix_to_prefix)
+    # copy package into place (shutil.copytree)
+    shutil.copytree(os.path.join(tempdir, spec.dag_hash()), spec.prefix,
+                    ignore=shutil.ignore_patterns('.spack/spec.json'))
```
Comment (vsoch) [inline lib/spack/spack/rewiring.py]: This is going to get angry if the prefix already exists - if it does and has the same hash I think we already have the spec and don't need to do again?

```suggestion
    if not os.path.exists(spec.prefix):
        shutil.copytree(os.path.join(tempdir, spec.dag_hash()), spec.prefix,
                        ignore=shutil.ignore_patterns('.spack/spec.json'))
```
but if it's the case that the relocations are different (and the hash is the same) then you'd instead want to do:

```suggestion
    if os.path.exists(spec.prefix):
        shutil.rmtree(spec.prefix)
    shutil.copytree(os.path.join(tempdir, spec.dag_hash()), spec.prefix,
                    ignore=shutil.ignore_patterns('.spack/spec.json'))
```

File: lib/spack/spack/rewiring.py [hunk 2]
```diff
@@ -37,7 +37,8 @@ def rewire(spliced_spec):
     assert spliced_spec.spliced
     for spec in spliced_spec.traverse(order='post', root=True):
         if not spec.build_spec.package.installed:
-            raise RuntimeError('Pure spec package was not installed')
+            # TODO: May want to change this at least for the root spec...
+            spec.build_spec.package.do_install(force=True)
```
Comment (vsoch) [inline lib/spack/spack/rewiring.py]: yep this is exactly what I did!

File: lib/spack/spack/rewiring.py (hunk 1 above)
Comment (becker33) [reply lib/spack/spack/rewiring.py]: If the relocations are different the hash has to be different too -- the hash contains all of the information on all of the splices.

File: lib/spack/spack/rewiring.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+import os
+import re
+import shutil
+import tempfile
+
+from ordereddict_backport import OrderedDict
```
Comment (becker33) [inline lib/spack/spack/rewiring.py]: You shouldn't need this anymore, since we're dropping python 2.6 support

File: lib/spack/spack/rewiring.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+
+def rewire(spliced_spec):
+    """Given a spliced spec, this function conducts all the rewiring on all
+    nodes in the DAG of that spec."""
+    assert spliced_spec.spliced
+    for spec in spliced_spec.traverse(order='post', root=True):
+        if not spec.build_spec.package.installed:
+            # TODO: May want to change this at least for the root spec...
+            spec.build_spec.package.do_install(force=True)
```
Comment (becker33) [inline lib/spack/spack/rewiring.py]: yikes... be careful what you ask for.

Probably reasonable, but someone could easily accidentally sign up for a lot of builds. I think we'll need some equivalent to `spack spec` that shows what will be built/rewired for a given command.

File: lib/spack/spack/rewiring.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+                                           str(spack.store.layout.root),
+                                           prefix_to_prefix,
+                                           False,
+                                           spec.build_spec.prefix,
+                                           spec.prefix)
+        relocate.relocate_text_bin(binaries=bins_to_relocate,
+                                   prefixes=prefix_to_prefix)
+    # copy package into place (shutil.copytree)
+    shutil.copytree(os.path.join(tempdir, spec.dag_hash()), spec.prefix,
+                    ignore=shutil.ignore_patterns('spec.json'))
```
Comment (becker33) [inline lib/spack/spack/rewiring.py]: This should also ignore the install manifest -- that will be recreated for the new install by the post install hooks.

File: lib/spack/spack/rewiring.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+    if manifest.get('link_to_relocate'):
+        _relocate_spliced_links(manifest.get('link_to_relocate'),
+                                spec.build_spec.prefix,
+                                spec.prefix)
+    shutil.rmtree(os.path.join(tempdir, spec.dag_hash()))
```
Comment (becker33) [inline lib/spack/spack/rewiring.py]: Shouldn't this remove `tmpdir` instead? It was created in this scope, I think this is a small inode leak

File: lib/spack/spack/test/rewiring.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+        for dep in node.traverse(root=True):
+            bin_file_path = os.path.join(dep.prefix.bin, bin_names[dep.name])
+            assert text_in_bin(dep.prefix, bin_file_path)
+        # test install manifest
+        assert spack.store.layout.check_installed(node)
+        manifest_file_path = os.path.join(node.prefix,
+                                          spack.store.layout.metadata_dir,
+                                          spack.store.layout.manifest_file_name)
+        assert os.path.exists(manifest_file_path)
+        # monkeypatch the modulefile_generation hook to an accumulator
```
Comment (becker33) [inline lib/spack/spack/test/rewiring.py]: Is this still TODO?

**Review by becker33 — CHANGES_REQUESTED** [review]

File: lib/spack/spack/rewiring.py (hunk 1 above)
Comment (nhanford) [reply lib/spack/spack/rewiring.py]: Yup sorry for not resolving this earlier. We wound up fixing the issue with the check to install missing specs.

File: lib/spack/spack/rewiring.py (hunk 3 above)
Comment (nhanford) [reply lib/spack/spack/rewiring.py]: w00t! Replaced with `collections`

File: lib/spack/spack/rewiring.py (hunk 4 above)
Comment (nhanford) [reply lib/spack/spack/rewiring.py]: This is a good point especially if we allow specs with different names.

File: lib/spack/spack/rewiring.py (hunk 5 above)
Comment (nhanford) [reply lib/spack/spack/rewiring.py]: Updated to also ignore `install_manifest.json`.

File: lib/spack/spack/rewiring.py (hunk 6 above)
Comment (nhanford) [reply lib/spack/spack/rewiring.py]: That was indeed an issue. Thanks for helping to root it out.

File: lib/spack/spack/test/rewiring.py (hunk 7 above)
Comment (nhanford) [reply lib/spack/spack/test/rewiring.py]: We decided this was low-priority enough and so unlikely to fail that we decided it wasn't critical to test.

File: lib/spack/spack/rewiring.py (hunk 4 above)
Comment (vsoch) [reply lib/spack/spack/rewiring.py]: This is exactly what we were talking about today! @becker33 to give you some idea, the current splicing says "Ok, here we have curl with zlib version A, let's splice in zlib version B. The implication of a current splice is "replace zlib A with zlib B" - we are splicing a different version/build of the "same" package. But let's say we have hdf5 which uses openmpi, and we want to splice in mpich - "replace openmpi with mpich entirely." It's a totally different package name and we'd want an API to do that too.

Or probably @nhanford said it better and shorter - we want to splice specs with different names :)

Comment (nhanford) [reply lib/spack/spack/rewiring.py]: Yeah and the concern is that we're going to risk some pretty severe rebuild events, so when I finally get around to the CLI, we want to make sure the user can see what they're about to trigger...

Comment (vsoch) [reply lib/spack/spack/rewiring.py]: Totally! Probably something similar to what spack solve does (what currently shows the result of `--reuse`) would make sense?

File: lib/spack/spack/rewiring.py (hunk 3 above)
Comment (becker33) [reply lib/spack/spack/rewiring.py]: As discussed, we'll leave this for later.

**Review by becker33 — DISMISSED** [review]

**Comment by vsoch** [comment]
Are we going to be able to splice different specs in? E.g., openmpi or mpich or vice versa - is that going to be another PR?

**Comment by nhanford** [comment]
> 

Sorry for missing this question: It will be in an upcoming PR.

File: lib/spack/spack/test/rewiring.py [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+
+
+@pytest.mark.parametrize('transitive', [True, False])
+def test_rewire(mock_fetch, install_mockery, transitive):
+    spec = Spec('splice-t^splice-h~foo').concretized()
+    dep = Spec('splice-h+foo').concretized()
+    spec.package.do_install(force=True)
+    dep.package.do_install(force=True)
+    spliced_spec = spec.splice(dep, transitive=transitive)
+    assert spec is not spliced_spec
```
Comment (becker33) [inline lib/spack/spack/test/rewiring.py]: Is this checking for something in particular?

File: lib/spack/spack/test/rewiring.py [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+
+    # test that it made it into the database
+    rec = spack.store.db.get_record(spliced_spec)
+    installed_in_db = rec.installed if rec else False
+    assert installed_in_db
+
+    # check the file in the prefix has the correct paths
+    for node in spliced_spec.traverse(root=True):
+        text_file_path = os.path.join(node.prefix, node.name)
+        text = ''
```
Comment (becker33) [inline lib/spack/spack/test/rewiring.py]: This line is unnecessary. Variables can escape their scope in python.

File: lib/spack/spack/test/rewiring.py [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+        with open(text_file_path, 'r') as f:
+            text = f.read()
+        for modded_spec in node.traverse(root=True):
+            assert modded_spec.prefix in text
+        # test install manifest
+        spack.store.layout.ensure_installed(node)
+        manifest_file_path = os.path.join(node.prefix,
+                                          spack.store.layout.metadata_dir,
+                                          spack.store.layout.manifest_file_name)
+        assert os.path.exists(manifest_file_path)
```
Comment (becker33) [inline lib/spack/spack/test/rewiring.py]: Is this just checking that the manifest file exists? Should we check that it reflects the new information?

File: lib/spack/spack/test/rewiring.py [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+    for node in spliced_spec.traverse(root=True):
+        for dep in node.traverse(root=True):
+            bin_file_path = os.path.join(dep.prefix.bin, bin_names[dep.name])
+            assert text_in_bin(dep.prefix, bin_file_path)
+        # test install manifest
+        spack.store.layout.ensure_installed(node)
+        manifest_file_path = os.path.join(node.prefix,
+                                          spack.store.layout.metadata_dir,
+                                          spack.store.layout.manifest_file_name)
+        assert os.path.exists(manifest_file_path)
```
Comment (becker33) [inline lib/spack/spack/test/rewiring.py]: This seems to be covered by the previous test?

File: lib/spack/spack/rewiring.py [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+        prefix_to_prefix[build_dep.prefix] = spec[build_dep.name].prefix
+
+    manifest = bindist.get_buildfile_manifest(spec.build_spec)
+    platform = spack.platforms.by_name(spec.platform)
+    if manifest.get('text_to_relocate'):
+        text_to_relocate = [os.path.join(tempdir, spec.dag_hash(), rel_path)
+                            for rel_path in manifest.get('text_to_relocate')]
+        relocate.relocate_text(files=text_to_relocate,
+                               prefixes=prefix_to_prefix)
+    if manifest.get('binary_to_relocate'):
```
Comment (becker33) [inline lib/spack/spack/rewiring.py:69]:
```suggestion
    if 'binary_to_relocate' in manifest:
```
is more performant

File: lib/spack/spack/rewiring.py [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+    spack.hooks.pre_install(spec)
+    # compute prefix-to-prefix for every node from the build spec to the spliced
+    # spec
+    prefix_to_prefix = OrderedDict({spec.build_spec.prefix: spec.prefix})
+    for build_dep in spec.build_spec.traverse(root=False):
+        prefix_to_prefix[build_dep.prefix] = spec[build_dep.name].prefix
+
+    manifest = bindist.get_buildfile_manifest(spec.build_spec)
+    platform = spack.platforms.by_name(spec.platform)
+    if manifest.get('text_to_relocate'):
```
Comment (becker33) [inline lib/spack/spack/rewiring.py:64]:
```suggestion
    if 'text_to_relocate' in manifest:
```

File: lib/spack/spack/rewiring.py [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+                                           False,
+                                           spec.build_spec.prefix,
+                                           spec.prefix)
+        relocate.relocate_text_bin(binaries=bins_to_relocate,
+                                   prefixes=prefix_to_prefix)
+    # copy package into place (shutil.copytree)
+    shutil.copytree(os.path.join(tempdir, spec.dag_hash()), spec.prefix,
+                    ignore=shutil.ignore_patterns('spec.json',
+                                                  'install_manifest.json'))
+    if manifest.get('link_to_relocate'):
```
Comment (becker33) [inline lib/spack/spack/rewiring.py:94]:
```suggestion
    if 'link_to_relocate' in manifest:
```

**Review by becker33 — CHANGES_REQUESTED** [review]

File: lib/spack/spack/rewiring.py (hunk 13 above)
Comment (becker33) [inline lib/spack/spack/rewiring.py]:
```suggestion
    text_to_relocate = manifest.get('text_to_relocate', None):
    if text_to_relocate:
```
Then you can use it inside the block wiithout needing a separate lookup

File: lib/spack/spack/test/rewiring.py (hunk 8 above)
Comment (nhanford) [reply lib/spack/spack/test/rewiring.py]: Removed per our discussion; this is already tested elsewhere.

File: lib/spack/spack/test/rewiring.py (hunk 9 above)
Comment (nhanford) [reply lib/spack/spack/test/rewiring.py]: Whoops. Removed.

File: lib/spack/spack/test/rewiring.py (hunk 10 above)
Comment (nhanford) [reply lib/spack/spack/test/rewiring.py]: A new test, `test_rewire_writes_new_metadata` has been added with the described check.

File: lib/spack/spack/test/rewiring.py (hunk 11 above)
Comment (nhanford) [reply lib/spack/spack/test/rewiring.py]: This duplicated code has been eliminated.

File: lib/spack/spack/rewiring.py (hunk 13 above)
Comment (nhanford) [reply lib/spack/spack/rewiring.py]: This has been changed accordingly in both the `text` and `binary` cases.

**Review by becker33 — APPROVED** [review]

**Pull request merged**